In [48]:
import numpy as np
import pandas as pd

In [49]:
data = pd.DataFrame({
    'child': ['A', 'B', 'C', 'D', 'E'],
    'age': [5, 6, 7, 8, 9],
    'height': [100, 105, 108, 112, 115]
})

data

child  age  height
0     A    5     100
1     B    6     105
2     C    7     108
3     D    8     112
4     E    9     115

In [50]:
age = data['age']
height = data['height']

mean_age = np.mean(age)
mean_height = np.mean(height)

print(f'mean_age = {mean_age}')
print(f'mean_height = {mean_height}')

denominator = np.sum((age - mean_age)**2)
numerator = np.sum((age - mean_age) * (height - mean_height))

print(f'numerator = {numerator}')
print(f'denominator = {denominator}')

mean_age = 7.0
mean_height = 108.0
numerator = 37.0
denominator = 10.0


In [51]:
m = numerator / denominator
b = mean_height - (m * mean_age)
m
b

np.float64(3.7)

np.float64(82.1)

Final Equation:$$Height = 3.7(Age) + 82.1$$

Prediction

For Age 10: $y = 3.7(10) + 82.1 =$ $119.1 \text{ cm}$

In [52]:
x = 10
y = m * x + b
y

np.float64(119.1)

In [53]:
for x in age:
  y = m * x + b
  print(y)

100.6
104.3
108.0
111.69999999999999
115.4


In [54]:
age
type(age)
height
type(height)

0    5
1    6
2    7
3    8
4    9
Name: age, dtype: int64

pandas.core.series.Series

0    100
1    105
2    108
3    112
4    115
Name: height, dtype: int64

pandas.core.series.Series

In [55]:
height_pred = m * age + b
height_pred

0    100.6
1    104.3
2    108.0
3    111.7
4    115.4
Name: age, dtype: float64

In [56]:
# Sum Squared Residual
ssr = np.sum((height - height_pred) ** 2)
ssr

np.float64(1.1000000000000085)

In [57]:
#  Mean Squared Error
np.sum((height - height_pred) ** 2) / len(age)

np.float64(0.2200000000000017)

In [58]:
# Root Mean Squared Error
np.sqrt((np.sum((height - height_pred) ** 2)) / len(age))

np.float64(0.46904157598234475)

Formula:$$R^2 = 1 - \frac{SS_{res}}{SS_{tot}}$$Where:

$SS_{res}$ (Residual Sum of Squares): $\sum (y_{true} - \hat{y})^2$ — The error our model makes.

$SS_{tot}$ (Total Sum of Squares): $\sum (y_{true} - \bar{y})^2$ — The variation in the data itself.

In [59]:
ss_res = np.sum((height - height_pred) ** 2)
ss_res
ss_tot = np.sum((height - mean_height) ** 2)
ss_tot

r2 = 1 - (ss_res / ss_tot)
r2

np.float64(1.1000000000000085)

np.float64(138.0)

np.float64(0.9920289855072463)